### PR TITLE
Add MedSAM Reimplementation

### PR DESCRIPTION
**Pull request by anwai98**

### CODE REVIEW

**Review by constantinpape — COMMENTED** [review]
This is duplicating a lot of functionality; making use of the flexible design we have already implemented it should be possible to do this in < 20 lines.

File: micro_sam/training/medsam_trainer.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import time
+
+import torch
+from torchvision.utils import make_grid
+
+import torch_em
+from torch_em.trainer.logger_base import TorchEmLogger
+
+
+class MedSAMTrainer(torch_em.trainer.DefaultTrainer):
```
[the remark this answers is not in the document]
Comment (anwai98) [reply micro_sam/training/medsam_trainer.py]: I've updated the trainer to a flexible design. which now is more or less no code changes at all. Thanks for the feedback!

File: micro_sam/training/sam_trainer.py [hunk 2]
```diff
@@ -216,12 +222,13 @@ def _compute_iterative_loss(self, batched_inputs, y_one_hot, num_subiter, multim
             iou_regression_loss += net_iou_regression_loss
             mean_model_iou += net_mean_model_iou
 
-            # Determine the next prompts based on current predictions.
-            with torch.no_grad():
-                # Get the mask and logit predictions corresponding to the predicted object
-                # (per actual object) with the best IOU.
-                masks, logits = self._get_best_masks(batched_outputs, batched_iou_predictions)
-                batched_inputs = self._update_prompts(batched_inputs, y_one_hot, masks, logits)
+            if num_subiter > 1:
```
Comment (constantinpape) [inline micro_sam/training/sam_trainer.py:225]: We could be more efficient here, and check if we still have a next subiteration to go for which we need the masks and updated prompts. Then normal training would also profit from this.
So something like 
```python
if i < (num_subiter - 1):
...
```
(I think that's correct, but please double check that it's not off by one!)

File: micro_sam/training/medsam_trainer.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class MedSAMTrainer(SamTrainer):
+    """Trainer class for replicating the trainer of MedSAM (https://arxiv.org/abs/2304.12306)
+    """
+    def __init__(
+        self,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
```
Comment (constantinpape) [inline micro_sam/training/medsam_trainer.py:11]: We should set `n_sub_iterations` to 1 here. It doesn't make sense to require it as a kwarg.

**Review by constantinpape — COMMENTED** [review]
Looks good overall, just a few minor comments.

File: micro_sam/training/sam_trainer.py (hunk 2 above)
Comment (anwai98) [reply micro_sam/training/sam_trainer.py]: Ahha yes, makes sense. Thanks for the mention!

File: micro_sam/training/medsam_trainer.py (hunk 3 above)
Comment (anwai98) [reply micro_sam/training/medsam_trainer.py]: Done. Fixed the `n_sub_iteration` and `mask_prob` values in the trainer. Thanks for the mention!

**Comment by anwai98** [comment]
Thanks for the feedback! This is GTG from my side. Let me know if you spot something else.

File: micro_sam/training/sam_trainer.py [hunk 4]
```diff
@@ -216,12 +222,14 @@ def _compute_iterative_loss(self, batched_inputs, y_one_hot, num_subiter, multim
             iou_regression_loss += net_iou_regression_loss
             mean_model_iou += net_mean_model_iou
 
-            # Determine the next prompts based on current predictions.
-            with torch.no_grad():
-                # Get the mask and logit predictions corresponding to the predicted object
-                # (per actual object) with the best IOU.
-                masks, logits = self._get_best_masks(batched_outputs, batched_iou_predictions)
-                batched_inputs = self._update_prompts(batched_inputs, y_one_hot, masks, logits)
+            if num_subiter > 1:  # Update prompts for more than one sub-iterations
```
[the remark this answers is not in the document]
Comment (anwai98) [reply micro_sam/training/sam_trainer.py]: Oh yeah true, overlooked it. Sorry! Thanks for the mention!

**Review by constantinpape — APPROVED** [review]
